### 아이템 9. try-finally 보다는 try-with-resource를 사용하라 

---  
# try with resource은 왜 사용할까? 

꼭 회수해야 하는 자원을 다룰때는 try-finally보다 try-with-resource를 사용해야한다. try-with-resource를 사용하면 코드가 분명해지고, 만들어지는 예외정보도 명확하다. 또한 정확하고 쉽게 자원을 회수할 수 있다. 

### try - finally(기존 전통방식)

try-finally 는 기존의 블록 단위 예외 처리 방법이다.  

원래는 try-catch-finally 블록으로 이루어진다.    

- try : 처음에 실행하고 싶은 블록
- catch : 예외 발생시 처리하고 싶은 블록
- finally : 예외 발생 여부에 상관없이 최종적으로 처리하고 싶은 블록
- 

1) 자원을 하나 사용할 경우 

```java
//자원이 하나인 경우
static String firstLineOfFile(String path) throws IOException {
    BufferedReader br = new BufferedReader(new FileReader(path));
    try {
        return br.readLine();
    } finally {
        br.close();
    }
}
```

자원이 하나라도 문제가 생길 수 있다. 

readLine 메서드가 기기의 물리적인 문제로 예외를 던진다면 readLine메서드가 예외를 던지고, 같은 이유로 close메서드도 실패할 것 이다. 

이런 경우 두 번째 예외가 첫번째 예외를 집어삼키게되어 스택추적 내역(에러 스택 트레이스)에 첫번째 예외에 관한 정보는 남지 않게 되어 실제 디버깅이 어렵다. 

2) 자원이 둘 이상인 경우

```java
//자원이 둘 이상인 경우
static void copy(String src, String dst) throws IOException {
    InputStream in = new FileInputStream(src);
    try {
        OutputStream out = new FileOutputStream(dst);
        try {
            byte[] buf = new byte[BUFFER_SIZE];
            int n;
            while ((n = in.read(buf)) >= 0)
                out.write(buf, 0, n);
        } finally {
            out.close();
    } finally {
        in.close();
    }
}
```

두번째의 경우 자원이 두개 이상이기 때문에 코드가 복잡하고 지저분하다.

기존 문제를 해결하기 위해서 자바 7부터 try-with-resources가 생겼다. 

## ****try-with-resources****

가독성과 유지보수 편의성을 위해 리소스 작업을 포함하는 코드들의 예외처리를 할 수 있도록 하는 문법이다. 

AutoCloseable이라는 인터페이스를 구현하고 있으며 try-with-resource 사용가능 하다. 

try-with-resource를 사용하면 try문이 종료될 때 자동으로 close를 호출해준다. 

( AutoCloseable는 close 하나 있는 인터페이스다) 

1) 자원이 하나일 경우 

```java
static String firstLineOfFile(String path) throws Exception {
    try (BufferedReader br = new BufferedReader (
        new FileReader(path))) {
            return br.readLine();
    }
}
```

2) 자원이 둘 이상일 경우 

```java
static void copy(String src, String det) throws IOException {
    try (InputStream in = new FileInputStream(src);
        OutputStream out = new FileOutputStream(dst)) {
        byte[] buf = new byte[BUFFER_SIZE];
        int n;
        while ((n == in.read(buf)) > = 0)
            out.write(buf, 0, n);
}
```

3) catch 절과 같이 쓰는 모습

```java
public class Item9 {
    static String firstLineOfFile(String path, String defaultVal) {
        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            return br.readLine();
        } catch (IOException e) {
            return defaultVal;
        }
    }
}
```

- 코드가 더 짧고 가독성이 좋으며 문제 진단에도 좋다.
- readLine 메서드와 close 메서드 둘 다 예외가 난다면,  close에서 발생한 예외는 숨겨지고 readLine에서 발생한 예외는 기록된다. try-finally처럼 close 예외에 덮여지지 않는다.
- 숨겨진 예외도 버려지는게 아니라 스택추적내역에 숨겨졌다(suppressed) 꼬리표를 달고 출력된다.
- 자바7에서 Throwable에 추가된 getSuppressed메서드를 이용하면 코드에도 가져올 수 있다.
- try-with-resource에서도 try문을 중첩하지 않고도 catch문을 통해 다수의 예외를 처리할 수 있다.

### ****Interface AutoCloseable && Method close****

```java
//Class FileInputStream 소스 코드 일부
public class FileInputStream extends InputStream

//Abstract Class InputStream 소스 코드 일부
abstract class InputStream implements Closeable

//Interface Closeable  소스 코드 일부
public interface Closeable extends AutoCloseable {
   public void close() throws IOException;
}

//Interface AutoCloseable 소스 코드 일부
public interface AutoCloseable {
   void close() throws Exception;
}
```

FileInputStream 에서는 Closeable을 구현하고 있다. 

Closeable의 경우에는 try-with-resource를 사용할 수 있게 해주는 AutoCloseable를 상속받아서 사용하고 있다. 따라서 try-with-resource에 필수적인 close Method가 각각의 자원에 관련된 객체에 맞게 구현되어 있다.
